src/osp/coordinates.h

```diff
@@ -0,0 +1,144 @@
+// at runtime, used as indices for CoordinateSpace::m_components
+struct CCompX { using datatype_t = spaceint_t; };
+struct CCompY { using datatype_t = spaceint_t; };
+struct CCompZ { using datatype_t = spaceint_t; };
+struct CCompSat { using datatype_t = Satellite; };
+
+template <typename CCOMP_T>
+using ViewCComp_t = Corrade::Containers::StridedArrayView1D<typename CCOMP_T::datatype_t>;
+
+template<typename CCOMP_T>
+constexpr ccomp_id_t ccomp_id() noexcept
+{
+    return ccomp_family_t::type<CCOMP_T>;
+}
+
+using CoordinateView_t = std::optional<Corrade::Containers::StridedArrayView1D<void>>;
+
+/**
+ * @brief Stores positions, velocities, and other related data for Satellites,
+ *        and exposes them through read-only strided array views.
+ *
+ * CoordinateSpaces must be managed by some external system. They can store any
+ * kind of coordinate data, such as Cartesian XYZ or orbital parameters.
+ *
+ * As part of ECS, this is a way to store common components in separate buffers.
+ * that can each be managed by a specific system (simultaneously too).
+ * aka: a non-EnTT but still ECS component pool
+ *
+ */
+struct CoordinateSpace
+{
+    using SatToAdd_t = std::tuple<Satellite, Vector3g, Vector3>;
+
+    enum class ECmdOp : uint8_t { Add, Set };
+    enum class ECmdVar : uint8_t { Position, Velocity };
+    using CmdValue_t = std::variant<Vector3, Vector3g>;
+
+    using Command_t = std::tuple<ECmdOp, ECmdVar, CmdValue_t>;
+
+    /**
+     * @brief Request to add a Satellite to this coordinate space
+     *
+     * Every coordinate space must be able to accept Satellites by a position
+     * and a velocity as a common interface. If the CoordinateSpace uses a
+     * non-Cartesian coordinate system like Kepler orbits, then it must be
+     * converted.
+     *
+     * @param sat [in] Satellite to add
+     * @param pos [in] Position in this coordinate system
+     * @param vel [in] Velocity of satellite
+     */
+    void add(Satellite sat, Vector3g pos, Vector3 vel)
+    {
+        m_toAdd.emplace_back(sat, pos, vel);
+    }
+
+    /**
+     * @brief TODO - commands to move and accelerate satellites
+     *
+     * @param cmd
+     */
+    void command(Command_t&& cmd)
+    {
+        m_commands.emplace_back(std::move(cmd));
+    }
+
+    /**
+     * @brief Access this CoordinateSpace's components
+     *
+     * Use the index from a Satellite's UCompCoordspaceIndex to access this view
+     *
+     * @return A StridedArrayView1D viewing coordinate space component data
+     *
+     * @tparam COMP_T Coordinate component to view
+     */
+    template<typename CCOMP_T>
+    ViewCComp_t<CCOMP_T> const ccomp_view() const
+    {
+        return Corrade::Containers::arrayCast<typename CCOMP_T::datatype_t>(
+                    m_components.at(ccomp_id<CCOMP_T>()).value());
+    }
+
+    // Queues for systems to communicate
+    std::vector<SatToAdd_t> m_toAdd;
+    std::vector<Satellite> m_toRemove;
+    std::vector<Command_t> m_commands;
```

Comment:
If these were instead std::lists, there would be a smaller cost for adding new items to them.
As well as the toAdd list being splicable into the destination.

Reply:
Small individually allocated POD types in std::lists really doesn't feel right. A reserved std::vector would outperform almost every case.

Considering concurrency, threads would need to push multiple elements into the queues all at once in its critical section; assuming this is why you're suggesting std::list. The best container I can think of is one of those lock-free atomic linked lists where each node is a block of multiple elements, so they aren't individually allocated.

Reply:
You're right that an atomic producer / consumer queue wouldn't be a bad call.

However, in this case, i was specifically concerned about the number of items these lists could have at a time.

You seem pretty satisfied that we're not going to get many at a time, so it's a mute point, and i agree that std::vector is a good choice.